// stanza/helpers/StreamManagement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const async_1 = require("async");
const events_1 = require("events");
const MAX_SEQ = Math.pow(2, 32);
const mod = (v, n) => ((v % n) + n) % n;
class StreamManagement extends events_1.EventEmitter {
    constructor() {
        super();
        this.allowResume = true;
        this.lastAck = 0;
        this.handled = 0;
        this.unacked = [];
        this.pendingRequest = false;
        this.inboundStarted = false;
        this.outboundStarted = false;
        this.id = undefined;
        this.jid = undefined;
        this.allowResume = true;
        this.started = false;
        this.cacheHandler = () => null;
        this._reset();
    }
    get started() {
        return this.outboundStarted && this.inboundStarted;
    }
    set started(value) {
        if (!value) {
            this.outboundStarted = false;
            this.inboundStarted = false;
        }
    }
    load(opts) {
        this.id = opts.id;
        this.allowResume = true;
        this.handled = opts.handled;
        this.lastAck = opts.lastAck;
        this.unacked = opts.unacked;
        this.emit('prebound', opts.jid);
    }
    cache(handler) {
        this.cacheHandler = handler;
    }
    bind(jid) {
        this.jid = jid;
        this._cache();
    }
    enable() {
        this.emit('send', {
            allowResumption: this.allowResume,
            type: 'enable'
        });
        this.handled = 0;
        this.outboundStarted = true;
    }
    resume() {
        this.emit('send', {
            handled: this.handled,
            previousSession: this.id,
            type: 'resume'
        });
        this.outboundStarted = true;
    }
    enabled(resp) {
        this.id = resp.id;
        this.handled = 0;
        this.inboundStarted = true;
        this._cache();
    }
    resumed(resp) {
        this.id = resp.previousSession;
        this.inboundStarted = true;
        this.process(resp, true);
        this._cache();
    }
    failed(resp) {
        // Resumption might fail, but the server can still tell us how far
        // the old session progressed.
        this.process(resp);
        // We alert that any remaining unacked stanzas failed to send. It has
        // been too long for auto-retrying these to be the right thing to do.
        for (const [kind, stanza] of this.unacked) {
            this.emit('failed', { kind, stanza });
        }
        this._reset();
        this._cache();
    }
    ack() {
        this.emit('send', {
            handled: this.handled,
            type: 'ack'
        });
    }
    request() {
        if (this.pendingRequest) {
            return;
        }
        this.pendingRequest = true;
        async_1.nextTick(() => {
            this.pendingRequest = false;
            this.emit('send', {
                type: 'request'
            });
        });
    }
    process(ack, resend = false) {
        if (ack.handled === undefined) {
            return;
        }
        const numAcked = mod(ack.handled - this.lastAck, MAX_SEQ);
        for (let i = 0; i < numAcked && this.unacked.length > 0; i++) {
            const [kind, stanza] = this.unacked.shift();
            this.emit('acked', { kind, stanza });
        }
        this.lastAck = ack.handled;
        if (resend) {
            const resendUnacked = this.unacked;
            this.unacked = [];
            for (const [kind, stanza] of resendUnacked) {
                this.emit('resend', { kind, stanza });
            }
        }
        this._cache();
    }
    track(kind, stanza) {
        if (kind !== 'message' && kind !== 'presence' && kind !== 'iq') {
            return;
        }
        if (this.outboundStarted) {
            this.unacked.push([kind, stanza]);
            this._cache();
            this.request();
        }
    }
    handle() {
        if (this.inboundStarted) {
            this.handled = mod(this.handled + 1, MAX_SEQ);
            this._cache();
        }
    }
    _cache() {
        this.cacheHandler({
            handled: this.handled,
            id: this.id,
            jid: this.jid,
            lastAck: this.lastAck,
            unacked: this.unacked
        });
    }
    _reset() {
        this.id = '';
        this.inboundStarted = false;
        this.outboundStarted = false;
        this.lastAck = 0;
        this.handled = 0;
        this.unacked = [];
    }
}
exports.default = StreamManagement;
